File: LeetCode/Valid_Palindrome.cpp
// Time Complexty: Linear, O(n) where n is the length of s
// Space Complexty: Constant, 0(1)

class Solution {
private:
    bool sameChar(char a, char b) {
        if (isalpha(a) and isalpha(b)) {
            return tolower(a) == tolower(b);
        }
        else {
            return a == b;
        }
    }

public:
    bool isPalindrome(string s) {
        int left = 0;
        int right = s.length()-1;

        while (left < right) {
            char leftChar = s[left];
            char rightChar = s[right];

            if (!isalnum(leftChar)) {
                left++;
                continue;
            }
            else if (!isalnum(rightChar)) {
                right--;
                continue;
            }
            else {
                if (!sameChar(leftChar, rightChar)) {
                    return false;
                }
            }
            left++;
            right--;
        }
        return true;
    }
};

File: LeetCode/Valid_Palindrome/Valid_Palindrome.py
# Time: O(n)
# Space: O(1)

class Solution:
    def validchar(self, c: chr) -> bool:
        return c.isalnum()

    def isPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s)-1
        s = s.lower()

        while left < right:
            if not self.validchar(s[left]):
                left += 1
            elif not self.validchar(s[right]):
                right -= 1
            else:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1

        return True
